### ISSUE TITLE
LiveSplit leaks user objects in memory, causing the app to crash when opening the context menu

### ISSUE THREAD

**Issue by SuperSajuuk** [LiveSplit/LiveSplit#2354]
Hi. I've been starting to use LiveSplit for making YouTube videos of older content, and so far, I find the app easy to use.... when actually timing stuff. The setting up part is much more frustrating, due primarily to the apps' constant penchant for randomly crashing for no reason at all.

So far, I've had crashes doing just a sample of the following:
- Editing split names by double clicking the split name
- Right clicking an empty part of the split window
- Opening the layout settings for a layout
- Pressing hotkeys that I've set up [eg the Reset hotkey]

There are other actions that I've found that cause a crash, but I don't remember which ones they are off-hand. The above are the most noticeable triggers for crashes, and there's no particular method that will always reproduce the crashes for them: after some time (which could be less than 60 seconds), one of the above will just cause the whole app to crash, often silently.

When I'm doing the first action -- editing split names -- the app complains about being unable to create a window handler: 
```
Error creating window handle.

   at System.Windows.Forms.NativeWindow.CreateHandle(CreateParams cp)
   at System.Windows.Forms.Control.CreateHandle()
   at System.Windows.Forms.Control.get_Handle()
   at System.Windows.Forms.Application.ParkHandle(CreateParams cp, DpiAwarenessContext dpiAwarenessContext)
   at System.Windows.Forms.Control.CreateHandle()
   at System.Windows.Forms.Control.get_Handle()
   at System.Windows.Forms.Control.CreateGraphicsInternal()
   at System.Windows.Forms.ThreadExceptionDialog..ctor(Exception t)
   at System.Windows.Forms.Application.ThreadContext.OnThreadException(Exception t)
   at System.Windows.Forms.Control.WndProcException(Exception e)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
   at System.Windows.Forms.UnsafeNativeMethods.DispatchMessageW(MSG& msg)
   at System.Windows.Forms.Application.ComponentManager.System.Windows.Forms.UnsafeNativeMethods.IMsoComponentManager.FPushMessageLoop(IntPtr dwComponentID, Int32 reason, Int32 pvLoopData)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoopInner(Int32 reason, ApplicationContext context)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoop(Int32 reason, ApplicationContext context)
   at LiveSplit.Program.Main(String[] args)
```
These crashes are extremely annoying because any of the changes that I have made before the crash are not saved. This means I have to repeat the process all over again and hope that the app won't just randomly crash.

I am running the app on Windows 10 Pro 22H2, Build 19045.3208. I have tried various fixes to stop the crashing, including running it as administrator [which I have to do in order for the global hotkeys to work in some of my games], but this has not helped.

EDIT: Done some digging, and there appears to be an occasional memory leak which causes user objects to rocket up to 10,000. See here for an easy set of reproduction steps: https://github.com/LiveSplit/LiveSplit/issues/2354#issuecomment-1742158945

Additionally, a method of stopping the crashes may be found in https://github.com/LiveSplit/LiveSplit/issues/2354#issuecomment-1742163603 which appears to involve opening and closing the app several times and using a hotkey to force the app to move to another monitor. Use Task Manager to make sure User Object count isn't running away, it should be idling around 70-80 in nominal circumstances.

**Comment by SuperSajuuk** [comment]
Cannot reproduce this any more. Suspecting this was an anti-virus issue, as I recently changed it and no longer suffer the crashes having done so.

**Comment by SuperSajuuk** [comment]
Nope, this has nothing to do with my anti-virus, unless every anti-virus manages to prevent this app working. Literally completed a run with this app, and as soon as I right-clicked the timing sheet to save splits, it crashed with this error.

Impossible to tell who the maintainers are of this repository, and from the looks of it, issue tickets tend to get no response, so I'm rather disappointed about the stability of this app. The many open issues with basically no responses suggests that this issue repository is being ignored and that the app is not really under development. Which is a shame imo.

**Comment by wooferzfg** [comment]
Make sure you have "Allow Gamepads as Hotkeys" disabled in the settings

**Comment by wooferzfg** [comment]
> Impossible to tell who the maintainers are of this repository, and from the looks of it, issue tickets tend to get no response, so I'm rather disappointed about the stability of this app. The many open issues with basically no responses suggests that this issue repository is being ignored and that the app is not really under development. Which is a shame imo.

This is just the reality with free software. My response to most issues would be "I have no idea what causes this", which isn't particularly helpful. I can't fix a bug when I don't know the cause and have no way to reproduce it - especially when it's related to complex interactions with drivers or low level Windows APIs. If you or someone else has a fix for it, I'd be happy to review a pull request.

**Comment by SuperSajuuk** [comment]
> Make sure you have "Allow Gamepads as Hotkeys" disabled in the settings
Thanks. I've checked and that is disabled. Global Hotkeys, Deactivate for Other Programs and Double Tap Prevention are enabled, but not the setting you mentioned, so I'm unclear as to why I get app crashes due to window handler failures.

**Comment by SuperSajuuk** [comment]
> Make sure you have "Allow Gamepads as Hotkeys" disabled in the settings

Hi @wooferzfg thanks for replying back. I've checked and that is disabled. Global Hotkeys, Deactivate for Other Programs and Double Tap Prevention are enabled, but not the setting you mentioned.

I have done some investigating and it appears this occurs due to a memory leak as described in this thread: https://stackoverflow.com/questions/222649/winforms-issue-error-creating-window-handle

LiveSplit does not appear to be properly cleaning up user objects, handles or threads. In my test, just right-clicking the window 2-3 times with a 2-3 second gap between attempts massively inflated User Objects to the 10,000 limit and crashed the app with a "Stopped Working" error. Presumably other silent CTD's are from similar memory leaks that hit the same limit in GDI Objects, Handles or Threads.

An easy set of reproduction steps to see this in action are:

1. open task manager
2. go to details tab
3. make sure handles, threads, user objects and gdi objects are displayed
4. open livesplit, find it in task manager
5. Right click the app to display the context menu. Spam-right clicks a couple of times.
6. Wait 2-3 seconds, then repeat 5 and 6.
7. While doing this, just observe Handles, Threads, User Objects and GDI Objects in the task manager, which will keep going up until it hits 10,000 and causes a crash.

Despite my message, I do appreciate the work that has gone in to making this app, but I just felt aggrieved when a 3.5 hour run doesn't get saved into the split times due to a random crash that seems to be outwith my control and has no solid reproduction.

**Comment by wooferzfg** [comment]
I've never been able to reproduce this issue. It appears to be similar to https://github.com/LiveSplit/LiveSplit/issues/2185

**Comment by SuperSajuuk** [comment]
Yeah looks like that issue to me, similar in set up to me in that we don't have the gamepads setting enabled [also checked that the settings.cfg has it disabled, which is the case]. Looks like perhaps there was a .NET update that has caused a memory leak or similar, might be worth trying to add something that just checks for whether the window handler error occurs and, if so, clears all user objects behind the scenes [but as I have no understanding of how to write .NET/C# code, I wouldn't know how this would be done].

**Comment by wooferzfg** [comment]
> Other crashes [which are silent] appears to be caused by differing errors relating to the .NET Runtime or even DirectInput. Quite why DirectInput would be causing crashes in LiveSplit mystifies me, considering that LiveSplit wouldn't have a reason to hook into a game controller

If you had the "Allow Gamepads as Hotkeys" setting off, then that gamepad code shouldn't have been getting called. You're not using any custom components or autosplitters? Have you already tried re-downloading LiveSplit and seeing if the issue still happens?

**Comment by SuperSajuuk** [comment]
Nope, the only components I've been using are the stock ones provided by the app. As for autosplitters, only using the ones that the app prompts me to use for the games I've been playiing so far. I did try re-downloading LiveSplit but that made no difference.

I'll remove the bit abouut the gamepads since that error doesn't appear to be relevant to the crash, which I've now just managed to stop occurring having done the stuff mentioned in my edit of my previous comment. The user objects column is no longer racing off to 10,000 in 60 seconds, and now stays within a nominal range of 70-100 with no apparent CTD's occurring.

**Comment by wooferzfg** [comment]
@SuperSajuuk Could you send me your splits and layout files?

**Comment by SuperSajuuk** [comment]
> @SuperSajuuk Could you send me your splits and layout files?

Hi @wooferzfg sorry for delay in replying. Is there a specific way you'd like these shared (ie Discord server or uploaded and linked here)? Happy to share this for you but need to know which route would be best. 👍

**Comment by wooferzfg** [comment]
@SuperSajuuk Whatever is easiest for you. (If you need to add me on Discord, my username is `@wooferzfg`)

**Comment by SuperSajuuk** [comment]
> @SuperSajuuk Whatever is easiest for you. (If you need to add me on Discord, my username is `@wooferzfg`)

@wooferzfg Hi there, thanks. I've sent you a friend request now, will share the details there once you've accepted. ^_^

**Comment by wooferzfg** [comment]
This ended up being an issue with an autosplitter component (`LiveSplit.UnrealLoads.dll`) being loaded even when not active, and that component initializing a `CompositeHook` that allowed gamepads as hotkeys, regardless of the user's choice in the settings. The fix for this was in #2366.
